### ISSUE TITLE
Can I run ComfyUI (with API) without having to open it on a browser? IS THAT POSSIBLE?

### ISSUE THREAD

**Issue by GPU-server** [comfyanonymous/ComfyUI#5433]
### Your question

I just closed my browser (I had launched 30 api calls that were running) and it immediately stoped and said "Processing interrupted".
I want to make comfy run (with api) even if it is not open on browser. Is that POSSIBLE?

**Comment by Amorano** [comment]
https://github.com/comfyanonymous/ComfyUI/blob/master/script_examples/websockets_api_example_ws_images.py

of course -- just use the API from whatever language you want -- use the example above for guidance.

**Comment by GPU-server** [comment]
> https://github.com/comfyanonymous/ComfyUI/blob/master/script_examples/websockets_api_example_ws_images.py
> 
> of course -- just use the API from whatever language you want -- use the example above for guidance.

I AM using the API!
When I closed the browser IT said: Processing interrupted.

**Comment by Amorano** [comment]
> > https://github.com/comfyanonymous/ComfyUI/blob/master/script_examples/websockets_api_example_ws_images.py
> > of course -- just use the API from whatever language you want -- use the example above for guidance.
> 
> I AM using the API! When I closed the browser IT said: Processing interrupted.

you said you closed your browser. The browser is using the API, yes, through the frontend code that you closed. It cant run the Queue if you closed the thing that is talking to the API.

When I say, "use the api" I mean the API -- not through a browser. If you want to control the Queue, you will need to use the actual API to make route calls to the specific routes to manage said Queue.

**Comment by ChristianKleineidam** [comment]
This shouldn't be closed. The server should continue to run in the background even if the UI closes if the server processes open jobs. I multiple times accidently aborted jobs by closing the UI. Jobs should continue to run without needing the UI to babysit them.

**Comment by NubeBuster** [comment]
> This shouldn't be closed. The server should continue to run in the background even if the UI closes if the server processes open jobs. I multiple times accidently aborted jobs by closing the UI. Jobs should continue to run without needing the UI to babysit them.

I believe this is just not in the scope of the design strategy that this project has. It's ComfyUI (read UI).

But worry not!
You can absolutely interface with the nodes using the api. There is an export (API) option in the ui web interface.


```bash
# 1. Start ComfyUI server (in one terminal)
cd ~/ComfyUI
python main.py

# 2. Run your workflow (in another terminal)
./run_workflow.sh /path/to/api_export.json
```

```bash
#!/bin/bash

# ComfyUI Headless Workflow Runner
# Usage: ./run_workflow.sh <workflow.json> [server_url]

set -e  # Exit on error

# Default settings
DEFAULT_SERVER="http://127.0.0.1:8188"
WORKFLOW_FILE="$1"
SERVER_URL="${2:-$DEFAULT_SERVER}"

# Colors for output
RED='\033[0;31m'
GREEN='\033[0;32m'
YELLOW='\033[1;33m'
BLUE='\033[0;34m'
NC='\033[0m' # No Color

# Helper functions
log_info() {
    echo -e "${BLUE}ℹ️  $1${NC}"
}

log_success() {
    echo -e "${GREEN}✅ $1${NC}"
}

log_warning() {
    echo -e "${YELLOW}⚠️  $1${NC}"
}

log_error() {
    echo -e "${RED}❌ $1${NC}"
}

# Usage function
usage() {
    echo "ComfyUI Headless Workflow Runner"
    echo ""
    echo "Usage: $0 <workflow.json> [server_url]"
    echo ""
    echo "Arguments:"
    echo "  workflow.json  Path to the ComfyUI workflow JSON file"
    echo "  server_url     ComfyUI server URL (default: $DEFAULT_SERVER)"
    echo ""
    echo "Examples:"
    echo "  $0 my_workflow.json"
    echo "  $0 my_workflow.json http://192.168.1.100:8188"
    echo ""
    echo "Note: Make sure ComfyUI server is running first:"
    echo "  python main.py --listen 127.0.0.1 --port 8188"
}

# Check arguments
if [ $# -eq 0 ] || [ "$1" = "-h" ] || [ "$1" = "--help" ]; then
    usage
    exit 0
fi

if [ ! -f "$WORKFLOW_FILE" ]; then
    log_error "Workflow file not found: $WORKFLOW_FILE"
    exit 1
fi

log_info "Starting ComfyUI workflow execution"
log_info "Workflow file: $WORKFLOW_FILE"
log_info "Server URL: $SERVER_URL"

# Test server connectivity
log_info "Testing server connectivity..."
if ! curl -s --connect-timeout 5 "$SERVER_URL/queue" > /dev/null; then
    log_error "Cannot connect to ComfyUI server at $SERVER_URL"
    log_error "Make sure ComfyUI is running with:"
    log_error "  python main.py --listen 127.0.0.1 --port 8188"
    exit 1
fi

log_success "Server is responding"

# Prepare the workflow file
TEMP_FILE=$(mktemp)
trap "rm -f $TEMP_FILE" EXIT

log_info "Preparing workflow data..."

# Check if the file already has "prompt" wrapper
if jq -e '.prompt' "$WORKFLOW_FILE" > /dev/null 2>&1; then
    log_info "File already has prompt wrapper"
    cp "$WORKFLOW_FILE" "$TEMP_FILE"
else
    log_info "Wrapping workflow in prompt structure"
    echo '{"prompt":' > "$TEMP_FILE"
    cat "$WORKFLOW_FILE" >> "$TEMP_FILE"
    echo '}' >> "$TEMP_FILE"
fi

# Validate JSON
if ! jq empty "$TEMP_FILE" 2>/dev/null; then
    log_error "Invalid JSON in workflow file"
    exit 1
fi

log_success "Workflow data prepared"

# Submit the workflow
log_info "Submitting workflow to ComfyUI..."

RESPONSE=$(curl -s -X POST "$SERVER_URL/prompt" \
    -H "Content-Type: application/json" \
    -d @"$TEMP_FILE")

# Check if submission was successful
if echo "$RESPONSE" | jq -e '.prompt_id' > /dev/null 2>&1; then
    PROMPT_ID=$(echo "$RESPONSE" | jq -r '.prompt_id')
    QUEUE_NUMBER=$(echo "$RESPONSE" | jq -r '.number')
    
    log_success "Workflow submitted successfully!"
    log_info "Prompt ID: $PROMPT_ID"
    log_info "Queue position: $QUEUE_NUMBER"
    
    # Check for node errors
    if echo "$RESPONSE" | jq -e '.node_errors' > /dev/null 2>&1; then
        NODE_ERRORS=$(echo "$RESPONSE" | jq '.node_errors')
        if [ "$NODE_ERRORS" != "null" ] && [ "$NODE_ERRORS" != "{}" ]; then
            log_warning "Workflow has node errors:"
            echo "$RESPONSE" | jq '.node_errors'
            log_warning "The workflow may still execute but might fail"
        fi
    fi
    
    # Monitor execution (optional)
    log_info "Monitoring execution progress..."
    echo "You can monitor progress at: $SERVER_URL"
    echo "Or check queue status with: curl -s $SERVER_URL/queue | jq"
    
    # Simple monitoring loop
    DOTS=""
    while true; do
        # Check if execution is complete
        HISTORY_RESPONSE=$(curl -s "$SERVER_URL/history/$PROMPT_ID")
        
        if echo "$HISTORY_RESPONSE" | jq -e ".[\"$PROMPT_ID\"]" > /dev/null 2>&1; then
            log_success "Workflow execution completed!"
            
            # Check for errors in history
            if echo "$HISTORY_RESPONSE" | jq -e ".[\"$PROMPT_ID\"].status.status_str" > /dev/null 2>&1; then
                STATUS=$(echo "$HISTORY_RESPONSE" | jq -r ".[\"$PROMPT_ID\"].status.status_str")
                if [ "$STATUS" = "success" ]; then
                    log_success "Execution status: $STATUS"
                else
                    log_error "Execution status: $STATUS"
                fi
            fi
            break
        fi
        
        # Show progress dots
        DOTS="${DOTS}."
        if [ ${#DOTS} -gt 10 ]; then
            DOTS=""
        fi
        printf "\r${BLUE}⏳ Executing${DOTS}${NC}"
        sleep 2
    done
    
    echo "" # New line after progress dots
    log_success "Workflow runner completed successfully!"
    
else
    log_error "Failed to submit workflow"
    log_error "Server response:"
    echo "$RESPONSE" | jq 2>/dev/null || echo "$RESPONSE"
    exit 1
fi
```

Note: untested code. But you can inspect the network tab in your browser to find the way to go. I couldn't get it to work with my strange workflows that have disable buttons.

**Comment by ChristianKleineidam** [comment]
> I believe this is just not in the scope of the design strategy that this project has. It's ComfyUI (read UI).

If I tell ComfyUI to create 50 images, why is the design philosophy that I shouldn't get my 50 images if I close the window after 20 were generated and the other 30 tasks are lost without ComfyUI even asking?

ComfyUI could have default (minimize to tray behavior) when the user clicks on close and there are open tasks.

**Comment by ltdrdata** [comment]
> > I believe this is just not in the scope of the design strategy that this project has. It's ComfyUI (read UI).
> 
> If I tell ComfyUI to create 50 images, why is the design philosophy that I shouldn't get my 50 images if I close the window after 20 were generated and the other 30 tasks are lost without ComfyUI even asking?
> 
> ComfyUI could have default (minimize to tray behavior) when the user clicks on close and there are open tasks.

Maybe you're looking for this feature: (yet merged)
https://github.com/Comfy-Org/desktop/pull/1206

This feature is only available in the Desktop version.

**Comment by NubeBuster** [comment]
> > I believe this is just not in the scope of the design strategy that this project has. It's ComfyUI (read UI).
> 
> If I tell ComfyUI to create 50 images, why is the design philosophy that I shouldn't get my 50 images if I close the window after 20 were generated and the other 30 tasks are lost without ComfyUI even asking?
> 
> ComfyUI could have default (minimize to tray behavior) when the user clicks on close and there are open tasks.

You're not wrong. This is the way the queue has been implemented. And it was a bit short-sighted perhaps. It must not be hard to make the queue server sided. Having thought about it, this is not a bad suggestion.

TL;DR I take back my statement that this is not in the design philosophy of ComfyUI, yet I emphasize that there is an inconvenient workaround.

**Comment by QuietNoise** [comment]
The queue **is** server side and it doesn't require ComfyUI to run in a browser. I never leave browser on for overnighth renders. 
Perhaps you have some extension that topples your processing.

**Comment by ChristianKleineidam** [comment]
@QuietNoise The issue is the Electron app and not using ComfyUI via a browser.

**Comment by QuietNoise** [comment]
I see. In that case yes, in the desktop app the server and front end are in one container. You can't separate then, it's by design and I don't think this will ever change as it is not a bug. Desktop app is a convenience product.
If you need separate instances for server and frontend they you run the server on it's own and connect to it from browser. Functionality wise you will get almost exact same experience plus the option to close browser and let the server run on its own.
